Data/ByteString.hs

```diff
@@ -1419,18 +1422,21 @@ findSubstring pat src
 findSubstrings :: ByteString -- ^ String to search for.
                -> ByteString -- ^ String to seach in.
                -> [Int]
-findSubstrings pat src
-    | null pat        = [0 .. ls]
-    | otherwise       = search 0
+findSubstrings pat = go
   where
-    lp = length pat
-    ls = length src
-    search !n
-        | (n > ls - lp) || null b = []
-        | otherwise = let k = n + length a
-                      in  k : search (k + lp)
+    breakPat = breakSubstring pat
+    go src
+        | null pat        = [0 .. ls]
+        | otherwise       = search 0
```

Comment:
Again, I think I'd advocate a style more like
```
findSubstrings pat =
    let breakPat = breakSubstring pat
     in \src -> ...
```
to 1. avoid having to give names like `go` to what are just non-recusrive function bodies, and 2. to make it blindly obvious that we're memoising something outside of a lambda. In the nested where style one has to look closely at the scoping to make sense of it.

Reply:
Again, that's fine, but it requires `MultiWayIf`